api/models/Activity.js

```diff
@@ -101,13 +136,86 @@ module.exports = bookshelf.Model.extend({
       reader_id: post.get('user_id'),
       actor_id: unfollowerId,
       post_id: post.id,
-      action: this.Action.Unfollow,
+      meta: {reasons: [this.Reason.Unfollow]},
       created_at: new Date()
     })
   },
 
   unreadCountForUser: function (user) {
     return Activity.query().where({reader_id: user.id, unread: true}).count()
     .then(rows => rows[0].count)
+  },
+
+  mergeReasons: function (reasons) {
+    const merged = values(reasons.reduce((acc, reason) => {
+      var current = acc[reason.reader_id]
+      if (acc[reason.reader_id]) {
+        const fields = ['actor_id', 'community_id']
+        fields.map(field => {
+          if (reason[field]) {
+            current[field] = reason[field]
+          }
+        })
+        current.reasons.push(reason.reason)
+      } else {
+        acc[reason.reader_id] = reason
+        acc[reason.reader_id].reasons = [acc[reason.reader_id].reason]
+      }
+      return acc
+    }, {}))
+    return merged
+  },
+
+  saveReasons: function (reasons, trx) {
```

Comment:
maybe `saveForReasons`? to emphasize that what's being saved is Activity instances